tests/widgets/test_menu.py

```diff
@@ -277,3 +298,87 @@ def test_data_choosers(asserter, qtbot, ij, gui_widget_chooser):
 
     # Assert that the data is now in napari
     asserter(lambda: 2 == len(button_to.viewer.layers))
+
+
+def test_settings_no_change(gui_widget: NapariImageJMenu):
+    """Ensure that no changes are made when there is no change from the defaults"""
+    button: SettingsButton = gui_widget.settings_button
+
+    # First record the old settings
+    old_yaml = [(k, v.get()) for k, v in settings.items()]
+
+    # Then update the settings, but select all defaults
+    button._update_settings()
+    # Then record the new settings and compare
+    new_yaml = [(k, v.get()) for k, v in settings.items()]
+    assert new_yaml == old_yaml
+
+
+def test_settings_change(asserter, gui_widget: NapariImageJMenu):
+    """Change imagej_directory_or_endpoint and ensure that the settings change"""
+    button: SettingsButton = gui_widget.settings_button
+
+    # REQUEST_VALUES MOCK
+    oldfunc = menu.request_values
+
+    key = "imagej_directory_or_endpoint"
+    old_value = settings[key]
+    new_value = "foo"
+
+    assert old_value != new_value
+
+    def newfunc(values={}, title="", **kwargs):
+        results = {}
+        # Solve each parameter
+        for name, options in values.items():
+            if name == "imagej_directory_or_endpoint":
+                results[name] = new_value
+                continue
+            elif "value" in options:
+                results[name] = options["value"]
+                continue
+
+            # Otherwise, we don't know how to solve that parameter
+            raise NotImplementedError()
+        return results
+
+    menu.request_values = newfunc
+
+    # # Start the handler in a new thread
+    class Handler(QRunnable):
+
+        # Test popup when running headlessly
+        def run(self) -> None:
+            asserter(lambda: isinstance(QApplication.activeWindow(), QMessageBox))
+            msg = QApplication.activeWindow()
+            expected_text = (
+                "Please restart napari for napari-imagej settings "
+                "changes to take effect!"
+            )
+            if expected_text != msg.text():
+                self._passed = False
+                return
+            ok_button = msg.button(QMessageBox.Ok)
+            ok_button.clicked.emit()
+            asserter(lambda: QApplication.activeModalWidget() is not msg)
+            self._passed = True
+
+        def passed(self) -> bool:
+            return self._passed
```

Comment:
Since this code is a near duplicate of the above `Handler`, maybe move this outside the function and allow the constructor to take some callbacks that allow us to check the contents of the `QMessageBox`?

Reply:
Cleaned up in 80d1733